src/core/ddsi/src/ddsi_raweth.c

```diff
@@ -109,6 +111,34 @@ static char *ddsi_raweth_to_string (char *dst, size_t sizeof_dst, const ddsi_loc
   return dst;
 }
 
+static void set_locator(ddsi_locator_t *srcloc, const uint8_t * addr, uint16_t port, uint16_t vtag)
+{
+  srcloc->kind = DDSI_LOCATOR_KIND_RAWETH;
+  srcloc->port = (uint32_t)(port +  ((vtag & 0xfff) << 20) + ((vtag & 0xf000) << 4));
```

Comment:
I'd prefer it if `port` and `vtag` were cast to a `uint32_t` before doing the arithmetic, so that all of it only involves unsigned 32-bit integers. I think the C99 section 6.5.7 "Bitwise shift operators" even says that if we are not careful we will get ourselves into trouble:

> The integer promotions are performed on each of the operands. The type of the result is that of the promoted left operand.
> The result of E1 << E2 is E1 left-shifted E2 bit positions; vacated bits are filled with zeros.
> If E1 has a signed type and nonnegative value, and E1 × 2^E2 is representable in the result type, then that is the resulting value; otherwise, the behavior is undefined.

So if `vtag & 0xfff` ≥ 0x800, so say 0x800 = 2^11, then `(vtag & 0xfff) << 20` = 2^11 × 2^20 = 2^31, and 2^31 isn't representable in a signed 32 bit int. That would make it undefined behaviour.

It is quite possible that this hasn't been introduced by the refactoring and that I just happened to spot it now. Even if that's the case, I think it makes sense to fix it now.

```suggestion
  srcloc->port = (uint32_t)port + (((uint32_t)vtag & 0xfff) << 20) + (((uint32_t)vtag & 0xf000) << 4));
```